cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -296,7 +297,15 @@ private boolean runReadyJob(Job job) throws Exception {
         return true;
       }
 
-      taskRunner.launch(job);
+      try {
+        taskRunner.launch(job);
+      } catch (TaskExecutionException e) {
```

Comment:
This can easily break as the code evolves if the taskRunner throws this exception for other reasons. You should check that the cause is a ProgramNotFoundException or ApplicationNotFoundException.